pkg/rollout/rollout.go

```diff
@@ -0,0 +1,191 @@
+package rollout
+
+import (
+	"fmt"
+
+	runapi "github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/run"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	format "github.com/logrusorgru/aurora"
+	"github.com/pkg/errors"
+	"google.golang.org/api/run/v1"
+)
+
+// Logger describes a logger printing information to the user.
+type Logger interface {
+	Println(args ...interface{})
+	Printf(format string, args ...interface{})
+}
+
+// Rollout is the rollout manager.
+type Rollout struct {
+	RunClient        runapi.Client
+	Config           *config.Config
+	Log              Logger
+	promoteCandidate bool
+}
+
+// Automatic tags.
+const (
+	StableTag    = "stable"
+	CandidateTag = "candidate"
+	LatestTag    = "latest"
+)
+
+// New returns a new rollout manager.
+func New(client runapi.Client, config *config.Config, log Logger) *Rollout {
+	return &Rollout{
+		RunClient:        client,
+		Config:           config,
+		Log:              log,
+		promoteCandidate: false,
+	}
+}
+
+// Manage handles the gradual rollout.
+func (r *Rollout) Manage() (*run.Service, error) {
+	serviceName := generateServiceName(r.Config.Metadata.Project, r.Config.Metadata.Service)
+	svc, err := r.RunClient.Service(serviceName)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get service")
+	}
+	if svc == nil {
+		return nil, errors.Errorf("service %s does not exist", serviceName)
+	}
+
+	stable := StableRevision(svc)
+	if stable == "" {
+		r.Log.Println("Could not determine stable revision")
+		return nil, nil
+	}
+
+	candidate := CandidateRevision(svc, stable)
+	if candidate == "" {
+		r.Log.Println("Could not determine candidate revision")
+		return nil, nil
+	}
+
+	svc = r.SplitTraffic(svc, stable, candidate)
+	svc = r.updateAnnotations(svc, stable, candidate)
+	svc, err = r.RunClient.ReplaceService(serviceName, svc)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not update service")
+	}
+
+	return svc, nil
+}
+
+// SplitTraffic changes the traffic configuration of the service.
+//
+// It creates a new traffic slice in case there's a new candidate revision.
+// If we only drop the traffic of the previous candidate to 0, Cloud Run would
+// still consider it as serving traffic and the slice of traffic targets would
+// grow very large over time.
+func (r *Rollout) SplitTraffic(svc *run.Service, stable, candidate string) *run.Service {
+	candidateTarget, candidatePercent := r.candidateTraffic(svc, candidate)
+	if candidateTarget == nil {
+		candidatePercent = r.Config.Rollout.Steps[0]
+	} else if candidatePercent == candidateTarget.Percent {
+		// If the traffic share did not change, candidate already handled 100%
+		// and is now ready to become stable.
+		r.promoteCandidate = true
+	}
+
+	var traffic []*run.TrafficTarget
```

Comment:
- do not try to modify spec.traffic section, just read it and gather info (candidate, stable, user-specified tags)

- create a new traffic section, with the candidate/stable rules (two rules)

- detect user-specified tags (rules without percentage, and tag != stable or candidate) `{revisionName: x, tag: y}`

- append the rules with user-specified tags into the new traffic section to preserve it